// datadog-lambda-js/dist/metrics/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.incrementBatchItemFailureMetric = exports.incrementInvocationsMetric = exports.incrementErrorsMetric = exports.KMSService = exports.MetricsQueue = exports.MetricsListener = void 0;
var listener_1 = require("./listener");
Object.defineProperty(exports, "MetricsListener", { enumerable: true, get: function () { return listener_1.MetricsListener; } });
var queue_1 = require("./queue");
Object.defineProperty(exports, "MetricsQueue", { enumerable: true, get: function () { return queue_1.MetricsQueue; } });
var kms_service_1 = require("./kms-service");
Object.defineProperty(exports, "KMSService", { enumerable: true, get: function () { return kms_service_1.KMSService; } });
var enhanced_metrics_1 = require("./enhanced-metrics");
Object.defineProperty(exports, "incrementErrorsMetric", { enumerable: true, get: function () { return enhanced_metrics_1.incrementErrorsMetric; } });
Object.defineProperty(exports, "incrementInvocationsMetric", { enumerable: true, get: function () { return enhanced_metrics_1.incrementInvocationsMetric; } });
Object.defineProperty(exports, "incrementBatchItemFailureMetric", { enumerable: true, get: function () { return enhanced_metrics_1.incrementBatchItemFailureMetric; } });
